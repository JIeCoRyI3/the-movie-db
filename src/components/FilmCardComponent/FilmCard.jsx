import React from 'react';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import styles from './FilmCard.module.css';
import { formatDate } from '../../utils/utils';
import { getSearchParams } from '../../utils/utils';

class FilmCard extends React.Component {
    state = {
        genres: null,
    };

    componentDidMount() {
        this.getGenres(this.props.genre_ids);
    }

    componentDidUpdate(prevProps) {
        if (prevProps.genres !== this.props.genres) {
            this.getGenres(this.props.genres);
        }
    }

    getGenres = (ids) => {
        const genres = this.props.genres
            .filter((genre) => ids.includes(genre.id))
            .map((genre) => genre.name);

        this.setState({
            genres: genres.join(', '),
        });
    };

    handleRoute = () => {
        if (this.props.id) {
            const urlParams = getSearchParams(this.props.location.search);
            const arrayOfParams = [];

            if (urlParams.searchBy) {
                arrayOfParams.push(`searchBy=${urlParams.searchBy}`);
            }
            if (urlParams.input) {
                arrayOfParams.push(`input=${urlParams.input}`);
            }
            if (urlParams.sortBy) {
                arrayOfParams.push(`sortBy=${urlParams.sortBy}`);
            }
            if (urlParams.sortType) {
                arrayOfParams.push(`sortType=${urlParams.sortType}`);
            }

            const getParams = arrayOfParams.join('&');
            this.props.history.push(`/film/${this.props.id}?${getParams}`);
            this.scrollToCurrentFilm();
        }
    };

    scrollToCurrentFilm = () => {
        document.documentElement.scrollTop = 0;
    };

    render() {
        return (
            <div
                onClick={this.handleRoute}
                className={`card ${styles.filmCard}`}
                style={{ width: 18 + 'rem' }}
            >
                <img
                    src={`https://image.tmdb.org/t/p/w500/${this.props.poster_path}`}
                    className={`card-img-top ${styles.filmCardImg}`}
                    alt="poster"
                />
                <div className={`card-body ${styles.filmCardBody}`}>
                    <h5 className={`card-title ${styles.filmCardTitle}`}>
                        {this.props.original_title}
                    </h5>

                    {!!this.props.release_date &&
                    formatDate(this.props.release_date) ? (
                        <p className={`card-text ${styles.filmCardYear}`}>
                            {formatDate(this.props.release_date)}
                        </p>
                    ) : null}

                    <p className={`card-text ${styles.filmCardGenresList}`}>
                        {this.state.genres}
                    </p>
                </div>
            </div>
        );
    }
}

FilmCard.propTypes = {
    poster_path: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    release_date: PropTypes.string.isRequired,
};

export default withRouter(FilmCard);
