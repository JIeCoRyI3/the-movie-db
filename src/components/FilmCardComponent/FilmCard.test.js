import React from 'react';
import { FilmCard } from './FilmCard';
import { shallow } from 'enzyme';

describe('<FilmCard>', () => {
    let wrapper;

    const props = {
        poster_path: 'poster',
        title: 'title text',
        release_date: '2323-20-20',
        genres: ['genres', 'genres', 'genres'],
        genre_ids: [1, 2, 3, 4],
    };

    beforeEach(() => {
        wrapper = shallow(<FilmCard {...props} />);
    });

    it('should render <FilmCard/> correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render image poster', () => {
        expect(wrapper.find('#filmCardImg')).toHaveLength(1);
    });

    it('should render film title', () => {
        expect(wrapper.find('#filmCardTitle')).toHaveLength(1);
    });

    // it('should render date', () => {
    //     expect(wrapper.find('#filmCardYear')).toHaveLength(1);
    // });

    it('should render genres', () => {
        expect(wrapper.find('#filmCardGenresList')).toHaveLength(1);
    });

    it('should handle click on card', () => {
        const instance = wrapper.instance();
        const mock = jest.fn();
        instance.props = { id: true, history: [mock()], location: {search: ''} };
        jest.spyOn(instance, 'handleRoute');
        wrapper.find('#filmCard').simulate('click');
        expect(mock).toHaveBeenCalled();
    });

    it('should handle error poster', () => {
        const instance = wrapper.instance();
        const mock = jest.fn();
        instance.props = { isPosterLoad: mock() };
        jest.spyOn(instance, 'errorPosterHandler');
        wrapper.find('#filmCardImg').simulate('error');
        expect(mock).toHaveBeenCalled();
    });
});
