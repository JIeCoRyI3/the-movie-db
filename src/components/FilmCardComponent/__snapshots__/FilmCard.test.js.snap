// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilmCard> should render <FilmCard/> correctly 1`] = `
<div
  className="card undefined"
  id="filmCard"
  onClick={[Function]}
  style={
    Object {
      "width": "18rem",
    }
  }
>
  <div>
    <img
      alt="Poster"
      className="card-img-top undefined"
      id="filmCardImg"
      onError={[Function]}
      src="https://image.tmdb.org/t/p/w500/poster"
    />
  </div>
  <div
    className="card-body undefined"
  >
    <div>
      <h5
        className="card-title undefined"
        id="filmCardTitle"
      />
    </div>
    <p
      className="card-text undefined"
      id="filmCardYear"
    >
      20.20.2323
    </p>
    <p
      className="card-text undefined"
      id="filmCardGenresList"
    />
  </div>
</div>
`;
